### ISSUE TITLE
incoming messages can timeout

### ISSUE THREAD

**Issue by yonahforst** [yonahforst/msg-gateway#2]
![screen shot 2017-01-12 at 1 10 07 pm](https://cloud.githubusercontent.com/assets/1440796/21889513/93c3fa46-d8c9-11e6-83c9-f19232acaa9a.png)

looks like incoming messages can timeout and throw an error. this can cause the sending service to retry. in our case, the chatbot ends up processing duplicate messages which totally fucks things up.

whatever is causing the process to timeout must be happening after the we've published to sns (because the chatbot is receiving the duplicates). The only thing that happens after is `logToAnalytics` so it's likely a call to one of the analytics services that's hanging.

this timeout isn't happening anymore, so it's hard to debug, but in any case we should be prepared to handle it in the future.

Probably the easiest solution is to use `setTimeout` with nodes `https.request` and set a value less than the lambda timeout. Lambda is timing out at 6000ms so maybe 3000? 

But then the problem becomes that we could potentially lose analytics events if the https request times out and we don't retry. So maybe we should somehow let the service retry, but suppress duplicate sns notifications? that's hard because we're completely stateless.

@behind-design any ideas?

**Comment by yonahforst** [comment]
In general we need to think about retries in this sns notification system, and how to prevent the parts that succeeded from trying.

**Comment by behind-design** [comment]
@yonahforst So basically the problem is that it looks like the analytics is failing and so the lambda function is timing out, and then it retries, and because of that we send duplicates to the user? Do I understand it correctly?

**Comment by yonahforst** [comment]
yes, thats exactly it. except these were inbound messages so we sent duplicate messages to the chatbot, causing confusing replies to the user

**Comment by behind-design** [comment]
@yonahforst I see that you're taking a look at the `test` stage. Do this happens only in it?

**Comment by yonahforst** [comment]
well by chance it was `test`, but if the problem was the call to the analytics service then it could happen anywhere.

maybe the solution, in general, is to push all non-critical work (logging and analytics across all our microservices) to an SQS queue. 

if we do that, we need a separate worker service to consume the queue, and we should probably remove analytics from the msg-gateway (we can still push the job to SQS from our own messageHandler)

**Comment by behind-design** [comment]
We can do that, or we can put it in the same way we're going to do the logging mechanism. That way we're centralising as well. But SQS is a good option too.

**Comment by yonahforst** [comment]
ok, lets start by adding `setTimeout`. better we miss analytics then confuse the user

**Comment by behind-design** [comment]
But with `setTimeout` I don't think we solve anything. Let me try first to figure out where the problem is, and if I don't find it I'll try to implement the `setTimeout`. Deal?

**Comment by yonahforst** [comment]
I would have expected setTimeout to throw an error if any request takes longer than 3s. then in the call to analytics, we can catch that error and still reply with a 200.

i just tested it by changing https.js like this
```js
\\...
  var mergedOptions = Object.assign({}, options, { headers, timeout: 3000 })

  return new Promise((resolve, reject) => {
    var req = https.request(mergedOptions, res => {
      res.setEncoding('utf8')
      res.setTimeout(3000, e => console.log('DID TIMEOUT', e))
\\...
```
and using network link conditioner to simulate a very slow network, but still got a timeout after 6 seconds.

**Comment by behind-design** [comment]
That's a good option, I'm thinking that if we were to move the logging to analytics to another lambda functions, and we trigger it, if that times out, it won't affect the msg-gateway one. Right?

**Comment by yonahforst** [comment]
@behind-design that's a great idea! so a third lambda function in the same sls config that subscribes to the outbound and inbound sns notifications, and handles logging. that's a much better solution, nice one!

**Comment by behind-design** [comment]
Sometimes I'd like to think I have good ideas.. :$
